virtcontainers/qemu.go

```diff
@@ -580,13 +588,77 @@ func (q *qemu) startSandbox(timeout int) error {
 		}
 	}()
 
+	if q.config.SharedFS == config.VirtioFS {
+		sockPath, err := q.vhostFSSocketPath(q.id)
+		if err != nil {
+			return err
+		}
+
+		// The daemon will terminate when the vhost-user socket
+		// connection with QEMU closes.  Therefore we do not keep track
+		// of this child process after returning from this function.
+		sourcePath := filepath.Join(kataHostSharedDir, q.id)
+		cmd := exec.Command(q.config.VirtioFSDaemon,
+			"-o", "vhost_user_socket="+sockPath,
+			"-o", "source="+sourcePath,
+			"-o", "cache="+q.config.VirtioFSCache)
+		stderr, err := cmd.StderrPipe()
+		if err != nil {
+			return err
+		}
+
+		if err = cmd.Start(); err != nil {
+			return err
+		}
+		defer func() {
+			if err != nil {
+				cmd.Process.Kill()
+			}
+		}()
+
+		// Wait for socket to become available
+		sockReady := make(chan error, 1)
+		timeStart := time.Now()
+		go func() {
+			scanner := bufio.NewScanner(stderr)
+			for scanner.Scan() {
+				if strings.Contains(scanner.Text(), "Waiting for QEMU socket connection...") {
```

Comment:
> This is really hacky and may block other virtiofsd implementations. One possible fix is to check the existence of the unix socket before and after `cmd.Start()`. For now, please at least add a `FIXME` above so that we can revisit it later.

Reporting readiness from a daemon is common practice.  For example, systemd has the sd_notify(3) interface where we would say "READY=1".

Polling for the existence of a UNIX domain socket has disadvantages:
1. A polling loop with a sleep() call adds latency.  The solution I've posted immediately detects that the daemon is ready without adding latency or busy loops.
2. There is a race condition between the daemon's bind(2)+listen(2) calls and QEMU's connect(2).  If QEMU is launched and attempts to connect between bind(2) and listen(2) the connect(2) call will fail.  This complicates things: either QEMU needs to add reconnect logic or kata-runtime needs to connect and pass the fd to QEMU, or the daemon needs to allow unexpected disconnects (from kata-runtime).

The only hacky thing I see here is the "QEMU" in the string we are checking.  virtiofsd works with other hypervisors and shouldn't know about QEMU.  I will change the output of virtiofsd so this string is generic.

Reply:
I've submitting a virtiofsd merge request to remove references to QEMU from the output:
https://gitlab.com/virtio-fs/qemu/merge_requests/4

Reply:
@stefanha The exact virtiofsd output string becomes part of the ABI we rely on for virtiofsd readiness, -- that is the part I'm not very comfortable with. Systemd's use of `sd_notify` is standard because `READY=1` etc state change is explicitly documented in man(3) sd_notify so it becomes part of the `sd_notify` API. If we decide to use the arbitrary string as virtiofsd readiness ABI, we need to declare it loud and clear so that any virtiofsd implementation knows it **MUST** print the exactly same string to stderr to announce its readiness. Right now it is hard-coded hidden in kata runtime code... To make it more acceptable, I think at least:
1. It needs to be defined as a static variable
2. We should document it explicitly somewhere obvious to virtiofsd implementers

OTOH, it seems that other vhost user daemons in `qemu/contrib/` do not report their readiness. How do they handle the issue in practice?

Reply:
Anyway, I'm OK to merge the PR as is since the feature is experimental. We can address the readiness issue (if necessary) in follow up PRs. @stefanha please fix the other comment and we can kick off the tests to see how CI says about it.

Reply:
I have pushed the latest version with GiB -> MiB fixed and QEMU -> vhost-user in the stderr message.

I will take a look at implementing sd_notify(3) in virtiofsd.  That way it will also work with systemd and we won't use stderr.  A separate file descriptor is passed for sd_notify(3) and that's a bit cleaner.  This will come as a separate pull request.

Reply:
Thanks @stefanha!